storage/s3/options.go

```diff
@@ -87,30 +221,13 @@ func NewOptions(opts ...*Options) *Options {
 		BufferSize: &defaultBufferSize,
 	}
 	for _, opt := range opts {
-		if opt.StaticCredentials != nil {
-			ret.StaticCredentials = opt.StaticCredentials
-		}
-		if opt.Region != nil {
-			ret.Region = opt.Region
+		ret.storageSettings.patch(&opt.storageSettings)
```

Comment:
```suggestion
		ret.storageSettings.patch(&opts[i].storageSettings)
```

Reply:
I'm not sure I understand the reasoning behind these two suggestions...

Reply:
I am paranoid, ever since I have realised that this:

```go
package main

import "fmt"

type Mode struct {
	number int
}

func main() {
	modes := []Mode{
		{
			number: 10,
		},
		{
			number: 11,
		},
	}
	for _, m := range modes {
		m.number = m.number * 2
	}
	fmt.Printf("modes:%+v", modes)
}
```

and this:
```go
package main

import "fmt"

type Mode struct {
	number int
}

func main() {
	modes := []Mode{
		{
			number: 10,
		},
		{
			number: 11,
		},
	}
	for i, _ := range modes {
		modes[i].number = modes[i].number * 2
	}
	fmt.Printf("modes:%+v", modes)
}
```

will actually print different results.

Reply:
That is correct, since the first example can be rewritten as:
```go
package main

import "fmt"

type Mode struct {
	number int
}

func main() {
	modes := []Mode{
		{
			number: 10,
		},
		{
			number: 11,
		},
	}
	for i := range modes {
		var m Mode = modes[i]
		m.number = m.number * 2
	}
	fmt.Printf("modes:%+v", modes)
}
```
However, since the iterable `opt` is read-only in this loop I find it cleaner to do this variable expansion inline.